### PR TITLE
[experimental] Introducing an enqueue state API

### PR DESCRIPTION
**Pull request by eczarny**
I wanted to explore an API that allows you to enqueue a state update that does not trigger a complete reflow of the component hierarchy. The idea here is to permit "lazy" state updates that aren't applied until a proper state update is scheduled.

There isn't anything here that can't be done without component controllers. Given that, and the somewhat confusing behavior of enqueued state updates, I don't recommend merging these changes. This pull request is completely experimental and for exploratory purposes only.

### CODE REVIEW

File: ComponentKit/HostingView/CKComponentHostingView.mm [hunk 1]
```diff
@@ -148,14 +148,23 @@ - (void)updateContext:(id<NSObject>)context mode:(CKUpdateMode)mode
...
 - (void)componentScopeHandleWithIdentifier:(CKComponentScopeHandleIdentifier)globalIdentifier
                             rootIdentifier:(CKComponentScopeRootIdentifier)rootIdentifier
-                     didReceiveStateUpdate:(id (^)(id))stateUpdate
+         didReceiveStateUpdateToBeEnqueued:(id (^)(id))updateBlock
                                       mode:(CKUpdateMode)mode
 {
   CKAssertMainThread();
-  _pendingInputs.stateUpdates.insert({globalIdentifier, stateUpdate});
-  [self _setNeedsUpdateWithMode:mode];
+  _pendingInputs.stateUpdates.insert({globalIdentifier, updateBlock});
```
Comment (eczarny) [inline ComponentKit/HostingView/CKComponentHostingView.mm]: This is naive and likely doesn't work, but it gets the idea across.

File: ComponentKit/TransactionalDataSources/Common/CKTransactionalComponentDataSource.mm [hunk 2]
```diff
@@ -157,11 +157,23 @@ - (void)componentScopeHandleWithIdentifier:(CKComponentScopeHandleIdentifier)glo
...
+{
+  CKAssertMainThread();
   if (mode == CKUpdateModeAsynchronous) {
-    _pendingAsynchronousStateUpdates[rootIdentifier].insert({globalIdentifier, stateUpdate});
+    _pendingAsynchronousStateUpdates[rootIdentifier].insert({globalIdentifier, updateBlock});
   } else {
-    _pendingSynchronousStateUpdates[rootIdentifier].insert({globalIdentifier, stateUpdate});
+    _pendingSynchronousStateUpdates[rootIdentifier].insert({globalIdentifier, updateBlock});
   }
 }
```
Comment (eczarny) [inline ComponentKit/TransactionalDataSources/Common/CKTransactionalComponentDataSource.mm]: This definitely doesn't work, you can verify that with my test in the sample app. The problem is this breaks the logic on line 155. Once you enqueue a state update the entire mechanism breaks. To make this work any enqueued updates would need to be tracked separately.

File: ComponentKit/Core/CKUpdateMode.h [hunk 3]
```diff
@@ -10,7 +10,14 @@
...
 typedef NS_ENUM(NSUInteger, CKUpdateMode) {
+  /** Apply the update off the main thread. */
   CKUpdateModeAsynchronous,
+  /**
+   Apply the update on the main thread.
+   The synchronous update mode will not immediately block the executing thread when apply updates. Instead ComponentKit
+   will schedule the update to be performed on the next tick of the main thread's run loop.
+   */
   CKUpdateModeSynchronous,
 };
```
Comment (ocrickard) [inline ComponentKit/Core/CKUpdateMode.h]: We should land these documentation changes.

Comment (eczarny) [inline ComponentKit/Core/CKUpdateMode.h]: That's why I kept the changes around. :) I'll get these changes into a proper pull request.

File: ComponentKit/TransactionalDataSources/Common/CKTransactionalComponentDataSource.mm (hunk 2 above)
Comment (ocrickard) [inline ComponentKit/TransactionalDataSources/Common/CKTransactionalComponentDataSource.mm]: lol yup, so simple.

**Comment by eczarny** [comment]
Closing since this should not be merged.

**Pull request closed without merging**